### PR TITLE
ctypes: remove use of legacy unicode API

### PR DESCRIPTION
**Pull request by methane**
PyUnicode_AsUnicodeAndSize() -> PyUnicode_AsWideChar()

### CODE REVIEW

**Review by auvipy — APPROVED** [review]

File: Modules/_ctypes/_ctypes.c [hunk 1]
```diff
@@ -1309,12 +1307,14 @@ WCharArray_set_value(CDataObject *self, PyObject *value, void *Py_UNUSED(ignored
...
-    wstr = PyUnicode_AsUnicodeAndSize(value, &len);
-    if (wstr == NULL)
+    Py_ssize_t len = PyUnicode_AsWideChar(value, NULL, 0);
+    if (len < 0) {
         return -1;
-    if ((size_t)len > self->b_size/sizeof(wchar_t)) {
-        PyErr_SetString(PyExc_ValueError,
-                        "string too long");
+    }
+    // PyUnicode_AsWideChar() returns number of wchars including trailing NUL,
```
Comment (brettcannon) [inline Modules/_ctypes/_ctypes.c]:
```suggestion
    // PyUnicode_AsWideChar() returns number of wchars including trailing null byte,
```

File: Modules/_ctypes/_ctypes.c [hunk 2]
```diff
@@ -1309,12 +1307,14 @@ WCharArray_set_value(CDataObject *self, PyObject *value, void *Py_UNUSED(ignored
...
+    // when it is called with NULL.
+    if ((size_t)len-1 > self->b_size/sizeof(wchar_t)) {
```
Comment (brettcannon) [inline Modules/_ctypes/_ctypes.c]:
```suggestion
    if (((size_t)len-1) > self->b_size/sizeof(wchar_t)) {
```

File: Modules/_ctypes/_ctypes.c [hunk 3]
```diff
@@ -1309,12 +1307,14 @@ WCharArray_set_value(CDataObject *self, PyObject *value, void *Py_UNUSED(ignored
...
+        PyErr_SetString(PyExc_ValueError, "string too long");
```
Comment (brettcannon) [inline Modules/_ctypes/_ctypes.c:1317]: Is it worth providing in the exception message the provided length and what the max length actually is?

File: Modules/_ctypes/cfield.c [hunk 4]
```diff
@@ -1242,9 +1239,14 @@ U_set(void *ptr, PyObject *value, Py_ssize_t length)
         return NULL;
     }
 
-    wstr = PyUnicode_AsUnicodeAndSize(value, &size);
-    if (wstr == NULL)
+    Py_ssize_t size = PyUnicode_AsWideChar(value, NULL, 0);
+    if (size < 0) {
         return NULL;
+    }
+    // PyUnicode_AsWideChar() returns number of wchars including trailing NUL,
```
Comment (brettcannon) [inline Modules/_ctypes/cfield.c]:
```suggestion
    // PyUnicode_AsWideChar() returns number of wchars including trailing null byte,
```

File: Modules/_ctypes/_ctypes.c (hunk 3 above)
Comment (methane) [reply Modules/_ctypes/_ctypes.c:1317]: Maybe yes.  But this file has similar codes which are not relating to wchar_t APIs. ("bytes string too long").
I want to keep this pull request simple.

**Pull request merged**